x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java

```diff
@@ -235,17 +254,19 @@ private RoleReferenceIntersection buildRoleReferencesForApiKey() {
     private RoleReferenceIntersection buildRoleReferencesForRemoteAccess() {
         final List<RoleReference> roleReferences = new ArrayList<>(4);
         @SuppressWarnings("unchecked")
-        final List<RemoteAccessAuthentication.RoleDescriptorsBytes> remoteAccessRoleDescriptorsBytes = (List<
-            RemoteAccessAuthentication.RoleDescriptorsBytes>) metadata.get(AuthenticationField.REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY);
+        final List<RoleDescriptorsBytes> remoteAccessRoleDescriptorsBytes = ((List<BytesReference>) metadata.get(
+            AuthenticationField.REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY
+        )).stream().map(b -> b instanceof RoleDescriptorsBytes ? (RoleDescriptorsBytes) b : new RoleDescriptorsBytes(b)).toList();
```

Comment:
This is a problem. When `RoleDescriptorsBytes` is deserialized, it is becomes a `BytesArray` because the methods `writeGenericMap` and `readMap` do not preserve sub-type information. The change here works. But it is non-ideal and I consider it a blocker.

The issue is similar to the discussion we had for the `authentication` object inside metadata. Reading/writing generic map cannot handle it and we had to store it as a string. Given the situation, I suggest we change how we serialise metadata, instead of completely handing it to generic map, providing finer control on how each member should be serialised, something like
```java
// Write
final Map<String, Writeable.Writer<?>> metadataValueWriter = Map.of(
    REMOTE_ACCESS_AUTHENTICATION_KEY, (o, v) -> ((Authentication) v).writeTo(o),
    REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY, (o, v) -> ((RoleDescriptorsBytes) v).writeTo(o),
    ...
);

out.writeVInt(metadata.size());
for (Map.Entry<String, Object> entry : metadata.entrySet()) {
    out.writeString(entry.getKey());
    @SuppressWarnings("unchecked")
    final var valueWriter = (Writeable.Writer<Object>) metadataValueWriter.getOrDefault(
        entry.getKey(),
        StreamOutput::writeGenericValue
    );
    valueWriter.write(out, entry.getValue());
}

// Read
final Map<String, CheckedFunction<StreamInput, Object, IOException>> metadataValueReader = Map.of(
    REMOTE_ACCESS_AUTHENTICATION_KEY, Authentication::new,
    REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY, RoleDescriptorsBytes::new,
    ...
);

final int metadataSize = in.readVInt();
final Map<String, Object> metadata = Maps.newMapWithExpectedSize(metadataSize);
for (int i = 0; i < metadataSize; i++) {
    final String key = in.readString();
    final Object value = metadataValueReader.getOrDefault(key, StreamInput::readGenericValue).apply(in);
    metadata.put(key, value);
}
```

Reply:
Thanks for catching this! Having gone back to the tests we have, we don't have a scenario where we call `buildRoleReferencesForRemoteAccess` after deserializing which why this didn't surface.

I think the approach you're proposing is sufficient. In the longer term, we might consider further moving away from storing authentication metadata as a map. That would be a bigger refactor though and as such something to tackle in the future. For now your solution is what we need.

Reply:
I agree it might be helpful to have metadata as its own dedicated class. But it's a lot cascading changes. For now I raised #93726 Thanks!